lib/classes/Variables.js

```diff
@@ -167,23 +168,37 @@ class Variables {
   }
 
   getValueFromSource(variableString) {
+    const cached = this.cache[variableString];
+    if (cached) {
+      return cached;
+    }
+    let valueFromSource;
     if (variableString.match(this.envRefSyntax)) {
-      return this.getValueFromEnv(variableString);
+      valueFromSource = this.getValueFromEnv(variableString);
     } else if (variableString.match(this.optRefSyntax)) {
-      return this.getValueFromOptions(variableString);
+      valueFromSource = this.getValueFromOptions(variableString);
     } else if (variableString.match(this.selfRefSyntax)) {
-      return this.getValueFromSelf(variableString);
+      valueFromSource = this.getValueFromSelf(variableString);
     } else if (variableString.match(this.fileRefSyntax)) {
-      return this.getValueFromFile(variableString);
+      valueFromSource = this.getValueFromFile(variableString);
     } else if (variableString.match(this.cfRefSyntax)) {
-      return this.getValueFromCf(variableString);
+      valueFromSource = this.getValueFromCf(variableString);
     } else if (variableString.match(this.s3RefSyntax)) {
-      return this.getValueFromS3(variableString);
+      valueFromSource = this.getValueFromS3(variableString);
     } else if (variableString.match(this.stringRefSynax)) {
-      return this.getValueFromString(variableString);
+      valueFromSource = this.getValueFromString(variableString);
     } else if (variableString.match(this.ssmRefSyntax)) {
-      return this.getValueFromSsm(variableString);
+      valueFromSource = this.getValueFromSsm(variableString);
+    }
+    if (valueFromSource) {
+      this.cache[variableString] = BbPromise.resolve(valueFromSource)
+      .then((result) => {
+        this.cache[variableString] = result
+        return result;
+      });
+      return valueFromSource;
```

Comment:
We should return the whole promise here - exactly the one that is stored in the cache map, so that the first invoker gets the same as all subsequent ones. This will guarantee consistent behavior.
```js
const resolver = BbPromise.resolve(valueFromSource)
  .then((result) => {
     this.cache[variableString] = result
     return result;
  });
this.cache[variableString] = resolver;
return resolver;
```

Reply:
Ah good catch! Thanks for the review feedback - I will fix it and eslint now.
I am really appreciative of how responsive and considered you and the serverless team are.